File: vermyth/bootstrap.py
from __future__ import annotations

from pathlib import Path

from vermyth.engine.composition import CompositionEngine
from vermyth.engine.projection_backends import ProjectionBackend, backend_from_env
from vermyth.engine.resonance import ResonanceEngine
from vermyth.grimoire.store import Grimoire
from vermyth.mcp.tools import VermythTools
from vermyth.observability import EventBus
from vermyth.registry import AspectRegistry
from vermyth.schema import ContradictionSeverity, EffectClass, Sigil


def build_tools(
    db_path: Path | str | None = None,
    *,
    backend: ProjectionBackend | None = None,
) -> tuple[Grimoire, CompositionEngine, ResonanceEngine, VermythTools]:
    """
    Build a fully wired Vermyth stack: Grimoire + composition + resonance + tools.

    Loads registered aspects and registered sigil overrides from the grimoire and
    applies them to the runtime registry and composition table.
    """
    grimoire = Grimoire(db_path=db_path)
    registry = AspectRegistry.get()
    latest_basis = grimoire.read_latest_basis_version()

    for aspect, _ordinal in grimoire.query_registered_aspects():
        registry.register(aspect)
    registry.set_basis_version(latest_basis.version)

    composition = CompositionEngine()
    for row in grimoire.query_registered_sigils():
        aspects = [registry.resolve(n) for n in row["aspects"]]
        aspects_fs = frozenset(aspects)
        sigil = Sigil(
            name=row["name"],
            aspects=aspects_fs,
            effect_class=EffectClass[row["effect_class"]],
            resonance_ceiling=float(row["resonance_ceiling"]),
            contradiction_severity=ContradictionSeverity[row["contradiction_severity"]],
        )
        entry = {
            "name": sigil.name,
            "aspects": [a.name for a in sigil.aspects],
            "effect_class": sigil.effect_class.name,
            "resonance_ceiling": sigil.resonance_ceiling,
            "contradiction_severity": sigil.contradiction_severity.name,
        }
        composition.register_sigil_entry(
            aspects_fs, entry, allow_override=bool(row.get("is_override", False))
        )

    engine = ResonanceEngine(
        composition_engine=composition,
        backend=backend,
        contradictions=composition.contradictions,
    )
    tools = VermythTools(engine, grimoire, event_bus=EventBus())
    return grimoire, composition, engine, tools


def build_tools_from_env(
    db_path: Path | str | None = None,
) -> tuple[Grimoire, CompositionEngine, ResonanceEngine, VermythTools]:
    """
    Convenience entry point for CLI/MCP. Uses env-driven backend selection.

    See `vermyth.engine.projection_backends.backend_from_env`.
    """

    backend = backend_from_env()
    return build_tools(db_path, backend=backend)

File: typescript/src/bootstrap.ts
import { CompositionEngine } from "./engine/composition.js";
import {
  NullProjectionBackend,
  backendFromEnv,
  type ProjectionBackend,
} from "./engine/projection.js";
import { ResonanceEngine } from "./engine/resonance.js";
import { Grimoire } from "./grimoire/store.js";
import { AspectRegistry } from "./registry.js";
import { aspectName } from "./schema/aspect.js";
import type { Aspect } from "./schema/aspect.js";
import { VermythTools } from "./mcp/tools/facade.js";
import { EventBus } from "./observability/event-bus.js";

export function buildTools(
  dbPath?: string | null,
  opts?: { backend?: ProjectionBackend | null },
): {
  grimoire: Grimoire;
  composition: CompositionEngine;
  engine: ResonanceEngine;
  tools: VermythTools;
} {
  const grimoire = new Grimoire(dbPath ?? undefined);
  const registry = AspectRegistry.get();
  const latest = grimoire.readLatestBasisVersion();

  for (const [aspect] of grimoire.queryRegisteredAspects()) {
    if (!registry.isRegistered(aspect.name)) {
      registry.register(aspect);
    }
  }
  registry.setBasisVersion(latest.version);

  const composition = new CompositionEngine();
  for (const row of grimoire.queryRegisteredSigils()) {
    const aspects = new Set<Aspect>();
    for (const n of row.aspects) {
      aspects.add(registry.resolve(n));
    }
    composition.registerSigilEntry(
      aspects,
      {
        name: row.name,
        aspects: [...aspects].map((a) => aspectName(a)),
        effect_class: row.effect_class,
        resonance_ceiling: row.resonance_ceiling,
        contradiction_severity: row.contradiction_severity,
      },
      { allowOverride: row.is_override },
    );
  }

  const engine = new ResonanceEngine(
    composition,
    opts?.backend ?? new NullProjectionBackend(),
    composition.contradictionsMap,
  );
  const tools = new VermythTools(engine, grimoire, new EventBus());
  return { grimoire, composition, engine, tools };
}

export function buildToolsFromEnv(dbPath?: string | null): {
  grimoire: Grimoire;
  composition: CompositionEngine;
  engine: ResonanceEngine;
  tools: VermythTools;
} {
  return buildTools(dbPath, { backend: backendFromEnv() });
}
